So the Linux job failed with...

```
/home/build-user/swift/lib/LLVMPasses/LLVMMergeFunctions.cpp:33:10: fatal error: 'clang/AST/StableHash.h' file not found
#include "clang/AST/StableHash.h"
         ^~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
```

--- a/clang/include/clang/Basic/LangOptions.def
+++ b/clang/include/clang/Basic/LangOptions.def
@@ -548,7 +548,7 @@ LANGOPT(BoundsAttributesCXXExperimental, 1, 0, "Experimental bounds attributes f
 LANGOPT(BoundsAttributesObjCExperimental, 1, 0, "Experimental bounds attributes for Objective-C")
 LANGOPT(BoundsSafetyRelaxedSystemHeaders, 1, 1,
         "Relax bounds safety assignment rules in system headers")
-ENUM_LANGOPT(BoundsSafetyBringUpMissingChecks, BoundsSafetyNewChecksMask, 7, BS_CHK_Default,
+LANGOPT(BoundsSafetyBringUpMissingChecks, 7, clang::LangOptions::getDefaultBoundsSafetyNewChecksMask(),
         "Bring up new -fbounds-safety run-time checks")
 /* TO_UPSTREAM(BoundsSafety) OFF*/
 

--- a/clang/include/clang/Basic/LangOptions.h
+++ b/clang/include/clang/Basic/LangOptions.h
@@ -447,7 +447,7 @@ public:
   };
 
   /* TODO(BoundsSafety) Deprecate the flag  */
-  enum BoundsSafetyNewChecksMask {
+  enum BoundsSafetyNewChecks {
     BS_CHK_None = 0,
 
     BS_CHK_AccessSize = 1 << 0,          // rdar://72252593
@@ -457,17 +457,13 @@ public:
     BS_CHK_CompoundLiteralInit = 1 << 4, // rdar://110871666
     BS_CHK_LibCAttributes = 1 << 5,      // rdar://84733153
     BS_CHK_ArraySubscriptAgg = 1 << 6,   // rdar://145020583
-
-    BS_CHK_All = BS_CHK_AccessSize | BS_CHK_IndirectCountUpdate |
-                 BS_CHK_ReturnSize | BS_CHK_EndedByLowerBound |
-                 BS_CHK_CompoundLiteralInit | BS_CHK_LibCAttributes |
-                 BS_CHK_ArraySubscriptAgg,
-
-    // This sets the default value assumed by clang if no
-    // `-fbounds-safety-bringup-missing-checks` flags are
-    // passed to clang.
-    BS_CHK_Default = BS_CHK_None,
   };
+  using BoundsSafetyNewChecksMaskIntTy =
+      std::underlying_type_t<BoundsSafetyNewChecks>;
+
+  static BoundsSafetyNewChecksMaskIntTy
+  getBoundsSafetyNewChecksMaskForGroup(StringRef GroupName);
+  static BoundsSafetyNewChecksMaskIntTy getDefaultBoundsSafetyNewChecksMask();
 
   /// Controls the various implementations for complex multiplication and
   // division.
@@ -897,10 +893,10 @@ public:
     return hasBoundsSafetyAttributes() && !hasBoundsSafety();
   }
 
-  // Take `enum BoundsSafetyNewChecksMask` value as input and return true if
-  // the mask is set. New -fbounds-safety run-time checks should use this helper
-  // to decide whether to enable/disable the checks.
-  bool hasNewBoundsSafetyCheck(BoundsSafetyNewChecksMask ChkMask) const {
+  // Take a BoundsSafetyNewChecksMask (or check mask) and return true if
+  // the check (or mask) is set. New -fbounds-safety run-time checks should use
+  // this helper to decide whether to enable/disable the checks.
+  bool hasNewBoundsSafetyCheck(BoundsSafetyNewChecksMaskIntTy ChkMask) const {
     return (BoundsSafetyBringUpMissingChecks & ChkMask) != 0;
   }
 

--- a/clang/include/clang/Driver/BoundsSafetyArgs.h
+++ b/clang/include/clang/Driver/BoundsSafetyArgs.h
@@ -13,10 +13,8 @@
 
 namespace clang {
 namespace driver {
-using BoundsSafetyNewChecksMaskTy =
-    std::underlying_type<LangOptions::BoundsSafetyNewChecksMask>::type;
 
-BoundsSafetyNewChecksMaskTy
+LangOptions::BoundsSafetyNewChecksMaskIntTy
 ParseBoundsSafetyNewChecksMaskFromArgs(const llvm::opt::ArgList &Args,
                                        DiagnosticsEngine *Diags);
 } // namespace driver

--- a/clang/include/clang/Driver/Options.td
+++ b/clang/include/clang/Driver/Options.td
@@ -1964,12 +1964,12 @@ defm bounds_safety_adoption_mode : BoolFOption<"bounds-safety-adoption-mode",
 def fbounds_safety_bringup_missing_checks_EQ : CommaJoined<["-"], "fbounds-safety-bringup-missing-checks=">, Group<f_Group>,
   MetaVarName<"<check>">,
   Visibility<[ClangOption, CC1Option]>,
-  HelpText<"Enable a set of new -fbounds-safety run-time checks, (option: access_size, indirect_count_update, return_size, ended_by_lower_bound, compound_literal_init, libc_attributes, array_subscript_agg, all)">;
+  HelpText<"Enable a set of new -fbounds-safety run-time checks, (option: access_size, indirect_count_update, return_size, ended_by_lower_bound, compound_literal_init, libc_attributes, array_subscript_agg, all, batch_0)">;
 
 def fno_bounds_safety_bringup_missing_checks_EQ : CommaJoined<["-"], "fno-bounds-safety-bringup-missing-checks=">, Group<f_Group>,
   MetaVarName<"<check>">,
   Visibility<[ClangOption, CC1Option]>,
-  HelpText<"Disable a set of new -fbounds-safety run-time checks, (option: access_size, indirect_count_update, return_size, ended_by_lower_bound, compound_literal_init, libc_attributes, array_subscript_agg, all)">;
+  HelpText<"Disable a set of new -fbounds-safety run-time checks, (option: access_size, indirect_count_update, return_size, ended_by_lower_bound, compound_literal_init, libc_attributes, array_subscript_agg, all, batch_0)">;
 
 def fbounds_safety_bringup_missing_checks : Flag<["-"], "fbounds-safety-bringup-missing-checks">, Group<f_Group>,
   Visibility<[ClangOption]>,

--- a/clang/lib/Basic/LangOptions.cpp
+++ b/clang/lib/Basic/LangOptions.cpp
@@ -240,3 +240,25 @@ LLVM_DUMP_METHOD void FPOptionsOverride::dump() {
 #include "clang/Basic/FPOptions.def"
   llvm::errs() << "\n";
 }
+
+/* TO_UPSTREAM(BoundsSafety) ON*/
+LangOptionsBase::BoundsSafetyNewChecksMaskIntTy
+LangOptionsBase::getBoundsSafetyNewChecksMaskForGroup(StringRef GroupName) {
+  const uint64_t Batch0 = BS_CHK_AccessSize | BS_CHK_IndirectCountUpdate |
+                          BS_CHK_ReturnSize | BS_CHK_EndedByLowerBound |
+                          BS_CHK_CompoundLiteralInit | BS_CHK_LibCAttributes |
+                          BS_CHK_ArraySubscriptAgg;
+
+  // Currently "all" and "batch_0" are the same
+  if (GroupName == "all" || GroupName == "batch_0")
+    return Batch0;
+
+  // Invalid GroupName
+  return BS_CHK_None;
+}
+
+LangOptionsBase::BoundsSafetyNewChecksMaskIntTy
+LangOptionsBase::getDefaultBoundsSafetyNewChecksMask() {
+  return BS_CHK_None;
+}
+/* TO_UPSTREAM(BoundsSafety) OFF*/

--- a/clang/lib/Driver/BoundsSafetyArgs.cpp
+++ b/clang/lib/Driver/BoundsSafetyArgs.cpp
@@ -16,7 +16,7 @@ namespace clang {
 
 namespace driver {
 
-BoundsSafetyNewChecksMaskTy
+LangOptions::BoundsSafetyNewChecksMaskIntTy
 ParseBoundsSafetyNewChecksMaskFromArgs(const llvm::opt::ArgList &Args,
                                        DiagnosticsEngine *Diags) {
   auto FilteredArgs =
@@ -24,22 +24,22 @@ ParseBoundsSafetyNewChecksMaskFromArgs(const llvm::opt::ArgList &Args,
                     OPT_fno_bounds_safety_bringup_missing_checks_EQ);
   if (FilteredArgs.empty()) {
     // No flags present. Use the default
-    return LangOptions::BS_CHK_Default;
+    return LangOptions::getDefaultBoundsSafetyNewChecksMask();
   }
 
   // If flags are present then start with BS_CHK_None as the initial mask and
   // update the mask based on the flags. This preserves compiler behavior for
   // users that adopted the `-fbounds-safety-bringup-missing-checks` flag when
-  // `BS_CHK_Default == BS_CHK_None`.
-  BoundsSafetyNewChecksMaskTy Result = LangOptions::BS_CHK_None;
+  // `getDefaultBoundsSafetyNewChecksMask() == BS_CHK_None`.
+  LangOptions::BoundsSafetyNewChecksMaskIntTy Result = LangOptions::BS_CHK_None;
   // All the options will be applied as masks in the command line order, to make
   // it easier to enable all but certain checks (or disable all but certain
   // checks).
   for (const Arg *A : FilteredArgs) {
     for (const char *Value : A->getValues()) {
-      std::optional<LangOptions::BoundsSafetyNewChecksMask> Mask =
+      std::optional<LangOptions::BoundsSafetyNewChecksMaskIntTy> Mask =
           llvm::StringSwitch<
-              std::optional<LangOptions::BoundsSafetyNewChecksMask>>(Value)
+              std::optional<LangOptions::BoundsSafetyNewChecksMaskIntTy>>(Value)
               .Case("access_size", LangOptions::BS_CHK_AccessSize)
               .Case("indirect_count_update",
                     LangOptions::BS_CHK_IndirectCountUpdate)
@@ -51,7 +51,11 @@ ParseBoundsSafetyNewChecksMaskFromArgs(const llvm::opt::ArgList &Args,
               .Case("libc_attributes", LangOptions::BS_CHK_LibCAttributes)
               .Case("array_subscript_agg",
                     LangOptions::BS_CHK_ArraySubscriptAgg)
-              .Case("all", LangOptions::BS_CHK_All)
+              .Case("all",
+                    LangOptions::getBoundsSafetyNewChecksMaskForGroup("all"))
+              .Case(
+                  "batch_0",
+                  LangOptions::getBoundsSafetyNewChecksMaskForGroup("batch_0"))
               .Case("none", LangOptions::BS_CHK_None)
               .Default(std::nullopt);
       if (!Mask) {

--- a/clang/lib/Frontend/CompilerInvocation.cpp
+++ b/clang/lib/Frontend/CompilerInvocation.cpp
@@ -4039,10 +4039,15 @@ void CompilerInvocationBase::GenerateLangArgs(const LangOptions &Opts,
 
   /*TO_UPSTREAM(BoundsSafety) ON*/
   if (Opts.BoundsSafety &&
-      Opts.BoundsSafetyBringUpMissingChecks != LangOptions::BS_CHK_Default) {
+      Opts.BoundsSafetyBringUpMissingChecks !=
+          LangOptions::getDefaultBoundsSafetyNewChecksMask()) {
     std::string Checks;
-    if (Opts.BoundsSafetyBringUpMissingChecks == LangOptions::BS_CHK_All) {
+    if (Opts.BoundsSafetyBringUpMissingChecks ==
+        LangOptions::getBoundsSafetyNewChecksMaskForGroup("all")) {
       Checks = "all";
+    } else if (Opts.BoundsSafetyBringUpMissingChecks ==
+               LangOptions::getBoundsSafetyNewChecksMaskForGroup("batch_0")) {
+      Checks = "batch_0";
     } else {
       if (Opts.hasNewBoundsSafetyCheck(LangOptions::BS_CHK_AccessSize))
         Checks += "access_size,";

--- a/clang/test/BoundsSafety/Driver/bounds-safety-bringup-missing-checks.c
+++ b/clang/test/BoundsSafety/Driver/bounds-safety-bringup-missing-checks.c
@@ -7,6 +7,8 @@
 // EMPTY-NOT: -fno-bounds-safety-bringup-missing-checks
 // EMPTY-NOT: -fbounds-safety-bringup-missing-checks=all
 // EMPTY-NOT: -fno-bounds-safety-bringup-missing-checks=all
+// EMPTY-NOT: -fbounds-safety-bringup-missing-checks=batch_0
+// EMPTY-NOT: -fno-bounds-safety-bringup-missing-checks=batch_0
 
 // RUN: %clang -fbounds-safety -fno-bounds-safety-bringup-missing-checks -c %s -### 2>&1 | FileCheck %s --check-prefix=DISABLED
 // DISABLED: -fno-bounds-safety-bringup-missing-checks=all
@@ -45,6 +47,9 @@
 // RUN: %clang -fbounds-safety -fbounds-safety-bringup-missing-checks=all -c %s -### 2>&1 | FileCheck %s --check-prefix=POS_ALL
 // POS_ALL: -fbounds-safety-bringup-missing-checks=all
 
+// RUN: %clang -fbounds-safety -fbounds-safety-bringup-missing-checks=batch_0 -c %s -### 2>&1 | FileCheck %s --check-prefix=POS_batch_0
+// POS_batch_0: -fbounds-safety-bringup-missing-checks=batch_0
+
 // RUN: %clang -fbounds-safety -fno-bounds-safety-bringup-missing-checks=access_size -c %s -### 2>&1 | FileCheck %s --check-prefix=NEG_ACCESS
 // NEG_ACCESS: -fno-bounds-safety-bringup-missing-checks=access_size
 
@@ -66,6 +71,9 @@
 // RUN: %clang -fbounds-safety -fno-bounds-safety-bringup-missing-checks=all -c %s -### 2>&1 | FileCheck %s --check-prefix=NEG_ALL
 // NEG_ALL: -fno-bounds-safety-bringup-missing-checks=all
 
+// RUN: %clang -fbounds-safety -fno-bounds-safety-bringup-missing-checks=batch_0 -c %s -### 2>&1 | FileCheck %s --check-prefix=NEG_batch_0
+// NEG_batch_0: -fno-bounds-safety-bringup-missing-checks=batch_0
+
 // RUN: %clang -fbounds-safety -fbounds-safety-bringup-missing-checks=access_size,indirect_count_update,return_size,ended_by_lower_bound,compound_literal_init,libc_attributes,all -c %s -### 2>&1 | FileCheck %s --check-prefix=POS_COMMA
 // POS_COMMA: -fbounds-safety-bringup-missing-checks=access_size,indirect_count_update,return_size,ended_by_lower_bound,compound_literal_init,libc_attributes,all
 
@@ -86,3 +94,8 @@
 // RUN: %clang -fbounds-safety-bringup-missing-checks -c %s -### 2>&1 | FileCheck %s --check-prefixes=UNUSED_POS
 // UNUSED_POS: warning: argument unused during compilation: '-fbounds-safety-bringup-missing-checks'
 
+// RUN: %clang -fno-bounds-safety-bringup-missing-checks=batch_0 -c %s -### 2>&1 | FileCheck %s --check-prefixes=UNUSED_NEG_BATCH_0
+// UNUSED_NEG_BATCH_0: warning: argument unused during compilation: '-fno-bounds-safety-bringup-missing-checks=batch_0'
+
+// RUN: %clang -fbounds-safety-bringup-missing-checks=batch_0 -c %s -### 2>&1 | FileCheck %s --check-prefixes=UNUSED_POS_BATCH_0
+// UNUSED_POS_BATCH_0: warning: argument unused during compilation: '-fbounds-safety-bringup-missing-checks=batch_0'

--- a/clang/unittests/Tooling/BoundsSafetyBringupMissingChecks.cpp
+++ b/clang/unittests/Tooling/BoundsSafetyBringupMissingChecks.cpp
@@ -131,45 +131,63 @@ bool runOnToolAndCheckLangOptions(const std::vector<std::string> &Args,
   return Result;
 }
 
-struct ChkPair {
+struct ChkDesc {
   const char *arg;
   unsigned Mask;
+  unsigned batch;
 };
 
-ChkPair Pairs[] = {
-    {"access_size", LangOptions::BS_CHK_AccessSize},
-    {"indirect_count_update", LangOptions::BS_CHK_IndirectCountUpdate},
-    {"return_size", LangOptions::BS_CHK_ReturnSize},
-    {"ended_by_lower_bound", LangOptions::BS_CHK_EndedByLowerBound},
-    {"compound_literal_init", LangOptions::BS_CHK_CompoundLiteralInit},
-    {"libc_attributes", LangOptions::BS_CHK_LibCAttributes},
-    {"array_subscript_agg", LangOptions::BS_CHK_ArraySubscriptAgg}};
-const size_t NumPairs = sizeof(Pairs) / sizeof(Pairs[0]);
-static_assert(NumPairs == 7, "Unexpected value");
+enum BatchedChecks { BATCH_ZERO = 0 };
+
+ChkDesc CheckKinds[] = {
+    {"access_size", LangOptions::BS_CHK_AccessSize, BATCH_ZERO},
+    {"indirect_count_update", LangOptions::BS_CHK_IndirectCountUpdate,
+     BATCH_ZERO},
+    {"return_size", LangOptions::BS_CHK_ReturnSize, BATCH_ZERO},
+    {"ended_by_lower_bound", LangOptions::BS_CHK_EndedByLowerBound, BATCH_ZERO},
+    {"compound_literal_init", LangOptions::BS_CHK_CompoundLiteralInit,
+     BATCH_ZERO},
+    {"libc_attributes", LangOptions::BS_CHK_LibCAttributes, BATCH_ZERO},
+    {"array_subscript_agg", LangOptions::BS_CHK_ArraySubscriptAgg, BATCH_ZERO}};
+const size_t NumChkDescs = sizeof(CheckKinds) / sizeof(CheckKinds[0]);
+static_assert(NumChkDescs == 7, "Unexpected value");
 
 // Check that `Pairs` is in sync with the `BoundsSafetyNewChecksMask`
-// enum.
+// enum and that the special "all" group contains all the checks
 TEST(BoundsSafetyBringUpMissingChecks, ChkPairInSync) {
   unsigned ComputedMask = 0;
   static_assert(LangOptions::BS_CHK_None == 0, "expected 0");
-  for (size_t Idx = 0; Idx < NumPairs; ++Idx) {
-    ComputedMask |= Pairs[Idx].Mask;
+  for (size_t Idx = 0; Idx < NumChkDescs; ++Idx) {
+    ComputedMask |= CheckKinds[Idx].Mask;
   }
-  ASSERT_EQ(ComputedMask, LangOptions::BS_CHK_All);
+  ASSERT_EQ(ComputedMask,
+            LangOptions::getBoundsSafetyNewChecksMaskForGroup("all"));
+}
+
+// Check that the "batch_0" group contains the right checks
+TEST(BoundsSafetyBringUpMissingChecks, ChkPairInSyncBatch0) {
+  unsigned ComputedMask = 0;
+  static_assert(LangOptions::BS_CHK_None == 0, "expected 0");
+  for (size_t Idx = 0; Idx < NumChkDescs; ++Idx) {
+    if (CheckKinds[Idx].batch == BATCH_ZERO)
+      ComputedMask |= CheckKinds[Idx].Mask;
+  }
+  ASSERT_EQ(ComputedMask,
+            LangOptions::getBoundsSafetyNewChecksMaskForGroup("batch_0"));
 }
 
 TEST(BoundsSafetyBringUpMissingChecks, ChkPairValidMask) {
   unsigned SeenBits = 0;
   static_assert(LangOptions::BS_CHK_None == 0, "expected 0");
-  for (size_t Idx = 0; Idx < NumPairs; ++Idx) {
-    unsigned CurrentMask = Pairs[Idx].Mask;
+  for (size_t Idx = 0; Idx < NumChkDescs; ++Idx) {
+    unsigned CurrentMask = CheckKinds[Idx].Mask;
     EXPECT_EQ(__builtin_popcount(CurrentMask), 1); // Check is a power of 2
     EXPECT_EQ(SeenBits & CurrentMask,
               0U); // Doesn't overlap with a previously seen value
     SeenBits |= CurrentMask;
     EXPECT_NE(SeenBits, 0U);
   }
-  ASSERT_EQ(SeenBits, LangOptions::BS_CHK_All);
+  ASSERT_EQ(SeenBits, LangOptions::getBoundsSafetyNewChecksMaskForGroup("all"));
 }
 
 #if 1
@@ -185,8 +203,8 @@ TEST(BoundsSafetyBringUpMissingChecks, ChkPairValidMask) {
 TEST(BoundsSafetyBringUpMissingChecks, DefaultWithBoundsSafety) {
   bool Result = runOnToolAndCheckLangOptions(
       {NEED_CC1_ARG "-fbounds-safety"}, [](LangOptions &LO) {
-        EXPECT_EQ(LO.getBoundsSafetyBringUpMissingChecks(),
-                  LangOptions::BS_CHK_Default);
+        EXPECT_EQ(LO.BoundsSafetyBringUpMissingChecks,
+                  LangOptions::getDefaultBoundsSafetyNewChecksMask());
       });
   ASSERT_TRUE(Result);
 }
@@ -194,13 +212,13 @@ TEST(BoundsSafetyBringUpMissingChecks, DefaultWithBoundsSafety) {
 TEST(BoundsSafetyBringUpMissingChecks, DefaultWithoutBoundsSafety) {
   bool Result = runOnToolAndCheckLangOptions(
       {NEED_CC1_ARG "-fno-bounds-safety"}, [](LangOptions &LO) {
-        EXPECT_EQ(LO.getBoundsSafetyBringUpMissingChecks(),
-                  LangOptions::BS_CHK_Default);
+        EXPECT_EQ(LO.BoundsSafetyBringUpMissingChecks,
+                  LangOptions::getDefaultBoundsSafetyNewChecksMask());
       });
   ASSERT_TRUE(Result);
   bool Result2 = runOnToolAndCheckLangOptions({""}, [](LangOptions &LO) {
-    EXPECT_EQ(LO.getBoundsSafetyBringUpMissingChecks(),
-              LangOptions::BS_CHK_Default);
+    EXPECT_EQ(LO.BoundsSafetyBringUpMissingChecks,
+              LangOptions::getDefaultBoundsSafetyNewChecksMask());
   });
   ASSERT_TRUE(Result2);
 }
@@ -210,8 +228,8 @@ TEST(BoundsSafetyBringUpMissingChecks,
   bool Result = runOnToolAndCheckLangOptions(
       {NEED_CC1_ARG "-fexperimental-bounds-safety-attributes"},
       [](LangOptions &LO) {
-        EXPECT_EQ(LO.getBoundsSafetyBringUpMissingChecks(),
-                  LangOptions::BS_CHK_Default);
+        EXPECT_EQ(LO.BoundsSafetyBringUpMissingChecks,
+                  LangOptions::getDefaultBoundsSafetyNewChecksMask());
       });
   ASSERT_TRUE(Result);
 }
@@ -225,12 +243,11 @@ TEST(BoundsSafetyBringUpMissingChecks, all_eq) {
       {NEED_CC1_ARG "-fbounds-safety",
        NEED_CC1_ARG "-fbounds-safety-bringup-missing-checks=all"},
       [](LangOptions &LO) {
-        EXPECT_EQ(LO.getBoundsSafetyBringUpMissingChecks(),
-                  LangOptions::BS_CHK_All);
+        EXPECT_EQ(LO.BoundsSafetyBringUpMissingChecks,
+                  LangOptions::getBoundsSafetyNewChecksMaskForGroup("all"));
 
-        for (size_t ChkIdx = 0; ChkIdx < NumPairs; ++ChkIdx) {
-          EXPECT_TRUE(LO.hasNewBoundsSafetyCheck(
-              (LangOptions::BoundsSafetyNewChecksMask)Pairs[ChkIdx].Mask));
+        for (size_t ChkIdx = 0; ChkIdx < NumChkDescs; ++ChkIdx) {
+          EXPECT_TRUE(LO.hasNewBoundsSafetyCheck(CheckKinds[ChkIdx].Mask));
         }
       });
   ASSERT_TRUE(Result);
@@ -241,12 +258,31 @@ TEST(BoundsSafetyBringUpMissingChecks, all) {
       {NEED_CC1_ARG "-fbounds-safety",
        NEED_CC1_ARG "-fbounds-safety-bringup-missing-checks"},
       [](LangOptions &LO) {
-        EXPECT_EQ(LO.getBoundsSafetyBringUpMissingChecks(),
-                  LangOptions::BS_CHK_All);
+        EXPECT_EQ(LO.BoundsSafetyBringUpMissingChecks,
+                  LangOptions::getBoundsSafetyNewChecksMaskForGroup("all"));
 
-        for (size_t ChkIdx = 0; ChkIdx < NumPairs; ++ChkIdx) {
-          EXPECT_TRUE(LO.hasNewBoundsSafetyCheck(
-              (LangOptions::BoundsSafetyNewChecksMask)Pairs[ChkIdx].Mask));
+        for (size_t ChkIdx = 0; ChkIdx < NumChkDescs; ++ChkIdx) {
+          EXPECT_TRUE(LO.hasNewBoundsSafetyCheck(CheckKinds[ChkIdx].Mask));
+        }
+      });
+  ASSERT_TRUE(Result);
+}
+
+// =============================================================================
+// batch_0 checks enabled
+// =============================================================================
+
+TEST(BoundsSafetyBringUpMissingChecks, batch_0) {
+  bool Result = runOnToolAndCheckLangOptions(
+      {NEED_CC1_ARG "-fbounds-safety",
+       NEED_CC1_ARG "-fbounds-safety-bringup-missing-checks=batch_0"},
+      [](LangOptions &LO) {
+        EXPECT_EQ(LO.BoundsSafetyBringUpMissingChecks,
+                  LangOptions::getBoundsSafetyNewChecksMaskForGroup("batch_0"));
+
+        for (size_t ChkIdx = 0; ChkIdx < NumChkDescs; ++ChkIdx) {
+          if (CheckKinds[ChkIdx].batch == BATCH_ZERO)
+            EXPECT_TRUE(LO.hasNewBoundsSafetyCheck(CheckKinds[ChkIdx].Mask));
         }
       });
   ASSERT_TRUE(Result);
@@ -257,25 +293,23 @@ TEST(BoundsSafetyBringUpMissingChecks, all) {
 // =============================================================================
 
 TEST(BoundsSafetyBringUpMissingChecks, only_one_check) {
-  for (size_t ChkIdx = 0; ChkIdx < NumPairs; ++ChkIdx) {
-    ChkPair Chk = Pairs[ChkIdx];
+  for (size_t ChkIdx = 0; ChkIdx < NumChkDescs; ++ChkIdx) {
+    ChkDesc Chk = CheckKinds[ChkIdx];
     std::vector<std::string> Args = {NEED_CC1_ARG "-fbounds-safety",
                                      NEED_CC1_ARG
                                      "-fbounds-safety-bringup-missing-checks="};
     Args[Args.size() - 1].append(Chk.arg);
 
     bool Result = runOnToolAndCheckLangOptions(Args, [&Chk](LangOptions &LO) {
-      EXPECT_EQ(LO.getBoundsSafetyBringUpMissingChecks(), Chk.Mask);
+      EXPECT_EQ(LO.BoundsSafetyBringUpMissingChecks, Chk.Mask);
 
       // Check helper
-      for (size_t OtherChkIdx = 0; OtherChkIdx < NumPairs; ++OtherChkIdx) {
-        ChkPair OtherChk = Pairs[OtherChkIdx];
+      for (size_t OtherChkIdx = 0; OtherChkIdx < NumChkDescs; ++OtherChkIdx) {
+        ChkDesc OtherChk = CheckKinds[OtherChkIdx];
         if (OtherChk.Mask == Chk.Mask) {
-          EXPECT_TRUE(LO.hasNewBoundsSafetyCheck(
-              (LangOptions::BoundsSafetyNewChecksMask)OtherChk.Mask));
+          EXPECT_TRUE(LO.hasNewBoundsSafetyCheck(OtherChk.Mask));
         } else {
-          EXPECT_FALSE(LO.hasNewBoundsSafetyCheck(
-              (LangOptions::BoundsSafetyNewChecksMask)OtherChk.Mask));
+          EXPECT_FALSE(LO.hasNewBoundsSafetyCheck(OtherChk.Mask));
         }
       }
     });
@@ -289,11 +323,12 @@ TEST(BoundsSafetyBringUpMissingChecks, only_one_check) {
 
 TEST(BoundsSafetyBringUpMissingChecks, all_pairs) {
   // Construct all distinct pairs and test
-  for (size_t firstIdx = 0; firstIdx < NumPairs; ++firstIdx) {
-    for (size_t secondIdx = firstIdx + 1; secondIdx < NumPairs; ++secondIdx) {
+  for (size_t firstIdx = 0; firstIdx < NumChkDescs; ++firstIdx) {
+    for (size_t secondIdx = firstIdx + 1; secondIdx < NumChkDescs;
+         ++secondIdx) {
       ASSERT_NE(firstIdx, secondIdx);
-      ChkPair First = Pairs[firstIdx];
-      auto Second = Pairs[secondIdx];
+      ChkDesc First = CheckKinds[firstIdx];
+      auto Second = CheckKinds[secondIdx];
       std::vector<std::string> Args = {
           NEED_CC1_ARG "-fbounds-safety",
           NEED_CC1_ARG "-fbounds-safety-bringup-missing-checks="};
@@ -305,17 +340,15 @@ TEST(BoundsSafetyBringUpMissingChecks, all_pairs) {
       bool Result = runOnToolAndCheckLangOptions(
           Args, [&First, &Second](LangOptions &LO) {
             unsigned ExpectedMask = First.Mask | Second.Mask;
-            EXPECT_EQ(LO.getBoundsSafetyBringUpMissingChecks(), ExpectedMask);
+            EXPECT_EQ(LO.BoundsSafetyBringUpMissingChecks, ExpectedMask);
 
             // Check helper
-            for (size_t ChkIdx = 0; ChkIdx < NumPairs; ++ChkIdx) {
-              unsigned ChkToTest = Pairs[ChkIdx].Mask;
+            for (size_t ChkIdx = 0; ChkIdx < NumChkDescs; ++ChkIdx) {
+              unsigned ChkToTest = CheckKinds[ChkIdx].Mask;
               if (ChkToTest == First.Mask || ChkToTest == Second.Mask) {
-                EXPECT_TRUE(LO.hasNewBoundsSafetyCheck(
-                    (LangOptions::BoundsSafetyNewChecksMask)ChkToTest));
+                EXPECT_TRUE(LO.hasNewBoundsSafetyCheck(ChkToTest));
               } else {
-                EXPECT_FALSE(LO.hasNewBoundsSafetyCheck(
-                    (LangOptions::BoundsSafetyNewChecksMask)ChkToTest));
+                EXPECT_FALSE(LO.hasNewBoundsSafetyCheck(ChkToTest));
               }
             }
           });
@@ -329,8 +362,8 @@ TEST(BoundsSafetyBringUpMissingChecks, all_pairs) {
 // =============================================================================
 
 TEST(BoundsSafetyBringUpMissingChecks, all_with_one_removed) {
-  for (size_t ChkIdx = 0; ChkIdx < NumPairs; ++ChkIdx) {
-    ChkPair Chk = Pairs[ChkIdx];
+  for (size_t ChkIdx = 0; ChkIdx < NumChkDescs; ++ChkIdx) {
+    ChkDesc Chk = CheckKinds[ChkIdx];
     std::vector<std::string> Args = {
         NEED_CC1_ARG "-fbounds-safety",
         NEED_CC1_ARG "-fbounds-safety-bringup-missing-checks=all",
@@ -338,18 +371,52 @@ TEST(BoundsSafetyBringUpMissingChecks, all_with_one_removed) {
     Args[Args.size() - 1].append(Chk.arg);
 
     bool Result = runOnToolAndCheckLangOptions(Args, [&Chk](LangOptions &LO) {
-      unsigned ExpectedMask = LangOptions::BS_CHK_All & (~Chk.Mask);
-      EXPECT_EQ(LO.getBoundsSafetyBringUpMissingChecks(), ExpectedMask);
+      unsigned ExpectedMask =
+          LangOptions::getBoundsSafetyNewChecksMaskForGroup("all") &
+          (~Chk.Mask);
+      EXPECT_EQ(LO.BoundsSafetyBringUpMissingChecks, ExpectedMask);
 
       // Check helper
-      for (size_t OtherChkIdx = 0; OtherChkIdx < NumPairs; ++OtherChkIdx) {
-        ChkPair OtherChk = Pairs[OtherChkIdx];
+      for (size_t OtherChkIdx = 0; OtherChkIdx < NumChkDescs; ++OtherChkIdx) {
+        ChkDesc OtherChk = CheckKinds[OtherChkIdx];
         if (OtherChk.Mask == Chk.Mask) {
-          EXPECT_FALSE(LO.hasNewBoundsSafetyCheck(
-              (LangOptions::BoundsSafetyNewChecksMask)OtherChk.Mask));
+          EXPECT_FALSE(LO.hasNewBoundsSafetyCheck(OtherChk.Mask));
         } else {
-          EXPECT_TRUE(LO.hasNewBoundsSafetyCheck(
-              (LangOptions::BoundsSafetyNewChecksMask)OtherChk.Mask));
+          EXPECT_TRUE(LO.hasNewBoundsSafetyCheck(OtherChk.Mask));
+        }
+      }
+    });
+    ASSERT_TRUE(Result);
+  }
+}
+
+TEST(BoundsSafetyBringUpMissingChecks, batch_0_with_one_removed) {
+  for (size_t ChkIdx = 0; ChkIdx < NumChkDescs; ++ChkIdx) {
+    ChkDesc Chk = CheckKinds[ChkIdx];
+    if (Chk.batch != 0)
+      continue;
+
+    std::vector<std::string> Args = {
+        NEED_CC1_ARG "-fbounds-safety",
+        NEED_CC1_ARG "-fbounds-safety-bringup-missing-checks=batch_0",
+        NEED_CC1_ARG "-fno-bounds-safety-bringup-missing-checks="};
+    Args[Args.size() - 1].append(Chk.arg);
+
+    bool Result = runOnToolAndCheckLangOptions(Args, [&Chk](LangOptions &LO) {
+      unsigned ExpectedMask =
+          LangOptions::getBoundsSafetyNewChecksMaskForGroup("batch_0") &
+          (~Chk.Mask);
+      EXPECT_EQ(LO.BoundsSafetyBringUpMissingChecks, ExpectedMask);
+
+      // Check helper
+      for (size_t OtherChkIdx = 0; OtherChkIdx < NumChkDescs; ++OtherChkIdx) {
+        ChkDesc OtherChk = CheckKinds[OtherChkIdx];
+        if (OtherChk.batch != BATCH_ZERO)
+          continue;
+        if (OtherChk.Mask == Chk.Mask) {
+          EXPECT_FALSE(LO.hasNewBoundsSafetyCheck(OtherChk.Mask));
+        } else {
+          EXPECT_TRUE(LO.hasNewBoundsSafetyCheck(OtherChk.Mask));
         }
       }
     });
@@ -366,7 +433,7 @@ TEST(BoundsSafetyBringUpMissingChecks, all_disabled) {
       {NEED_CC1_ARG "-fbounds-safety",
        NEED_CC1_ARG "-fno-bounds-safety-bringup-missing-checks"},
       [](LangOptions &LO) {
-        EXPECT_EQ(LO.getBoundsSafetyBringUpMissingChecks(),
+        EXPECT_EQ(LO.BoundsSafetyBringUpMissingChecks,
                   LangOptions::BS_CHK_None);
       });
   ASSERT_TRUE(Result);
@@ -378,7 +445,7 @@ TEST(BoundsSafetyBringUpMissingChecks, all_enable_then_disable) {
        NEED_CC1_ARG "-fbounds-safety-bringup-missing-checks",
        NEED_CC1_ARG "-fno-bounds-safety-bringup-missing-checks"},
       [](LangOptions &LO) {
-        EXPECT_EQ(LO.getBoundsSafetyBringUpMissingChecks(),
+        EXPECT_EQ(LO.BoundsSafetyBringUpMissingChecks,
                   LangOptions::BS_CHK_None);
       });
   ASSERT_TRUE(Result);
@@ -389,8 +456,8 @@ TEST(BoundsSafetyBringUpMissingChecks, all_enable_then_disable) {
 // =============================================================================
 
 TEST(BoundsSafetyBringUpMissingChecks, all_disabled_then_enable_one) {
-  for (size_t ChkIdx = 0; ChkIdx < NumPairs; ++ChkIdx) {
-    ChkPair Chk = Pairs[ChkIdx];
+  for (size_t ChkIdx = 0; ChkIdx < NumChkDescs; ++ChkIdx) {
+    ChkDesc Chk = CheckKinds[ChkIdx];
     std::vector<std::string> Args = {
         NEED_CC1_ARG "-fbounds-safety",
         NEED_CC1_ARG "-fno-bounds-safety-bringup-missing-checks",
@@ -398,17 +465,15 @@ TEST(BoundsSafetyBringUpMissingChecks, all_disabled_then_enable_one) {
     Args[Args.size() - 1].append(Chk.arg);
 
     bool Result = runOnToolAndCheckLangOptions(Args, [&Chk](LangOptions &LO) {
-      EXPECT_EQ(LO.getBoundsSafetyBringUpMissingChecks(), Chk.Mask);
+      EXPECT_EQ(LO.BoundsSafetyBringUpMissingChecks, Chk.Mask);
 
       // Check helper
-      for (size_t OtherChkIdx = 0; OtherChkIdx < NumPairs; ++OtherChkIdx) {
-        ChkPair OtherChk = Pairs[OtherChkIdx];
+      for (size_t OtherChkIdx = 0; OtherChkIdx < NumChkDescs; ++OtherChkIdx) {
+        ChkDesc OtherChk = CheckKinds[OtherChkIdx];
         if (OtherChk.Mask == Chk.Mask) {
-          EXPECT_TRUE(LO.hasNewBoundsSafetyCheck(
-              (LangOptions::BoundsSafetyNewChecksMask)OtherChk.Mask));
+          EXPECT_TRUE(LO.hasNewBoundsSafetyCheck(OtherChk.Mask));
         } else {
-          EXPECT_FALSE(LO.hasNewBoundsSafetyCheck(
-              (LangOptions::BoundsSafetyNewChecksMask)OtherChk.Mask));
+          EXPECT_FALSE(LO.hasNewBoundsSafetyCheck(OtherChk.Mask));
         }
       }
     });

